### PR TITLE
Ag gtsummary3

### PR DESCRIPTION
**Pull request by amygimma**
Still not finished!

But progress:
Done: gtsummary CFR and attack rate functions with add_stat and easy wrapper functions for beginners (just have to update some of the roxygen documentation)
In progress: gt univariate wrapper (need to have different outputs for OR/RR)
Todo: MH odds

### CODE REVIEW

**Comment by aspina7** [comment]
hey hey looks good - couple comments from just briefly going through. 

- [ ] I dont think we should export the functions to pass to `add_stat()`, would only export the user facing functions and name them e.g. add_cfr, add_ar and add_crosstab? 
- [ ] `gtsummary_case_fatality_rate()` is not happy with a yes/no variable (see **code snippet 1**  _example 1_ below) - guessing because is dichotomous but dont know which level to take? 
- [ ]  `gtsummary_case_fatality_rate()` returns the original died column from the tbl_summary (see **code snippet 1**  _example 2_ below) - can we automatically include `gt_remove_stat()` within the function so that users dont need to add it afterwards? 
- [ ] need to add cfr by strata - currently throws error "cfr by strata is not currently available, ignoring `by` argument"
- [ ] `gtsummary_attack_rate() `not cooperating (see **code snippet 1**  _example 3_ below)
- [ ] need to add attack rate by strata - same error as above
- [ ] for `merge_gt_univar_counts()` not entirely sure its possible to run a proper poisson regression with offset using `gtsummary::tbl_uvregression()` - cross reference with [epiet case studies](https://github.com/EPIET/MultivariableAnalysis/tree/master/MVA%202018) if unsure how these should be run appropriately


**code snippet 1** 
```
library("dplyr")
  library("gtsummary")
  
  ## load data 
  linelist <- epidict::gen_data("cholera")
  
  ###### EXAMPLE 1 
  ## make a character yes no var
  linelist$died <- ifelse(linelist$exit_status %in% c("DH", "DD", "DOA"), "yes", "no")
  
  ## try running function
  linelist %>% select(died) %>% tbl_summary() %>% gtsummary_case_fatality_rate()
  ## I get: 
  ## There was an error for variable 'died':
  ## Error: Problem with `summarise()` column `deaths`.
  ## i `deaths = sum(data[[variable]], na.rm = TRUE)`.
  ## x ungültiger 'type' (character) des Argumentes
  
  ####### EXAMPLE 2 
  ## whereas with a TRUE/FALSE variable it works 
  linelist$died <- linelist$died == "yes"
  linelist %>% select(died) %>% tbl_summary() %>% gtsummary_case_fatality_rate()

  ######## EXAMPLE 3  

  ## get count of number dead (different each time generate dataset) 
  num_dead <- sum(linelist$died, na.rm = TRUE)
  
  ## run epikit attack rate for reference 
  epikit::attack_rate(num_dead, 300)
  
  ## compare with gtsummary add_ci (not a good option as cant define our own denominator)
  linelist %>% 
     select(died) %>% 
     tbl_summary() %>% add_ci()
     
   ## compare with the wrapper (comes out with an AR of 100 - which is wrong...) 
   linelist %>% 
    select(died) %>% 
    tbl_summary() %>% 
    gtsummary_attack_rate(population = 300, 100)

```

**Comment by amygimma** [comment]
See responses below - can jump on a call tomorrow? 
Today I am updating documentation to be more clear and only exporting necessary functions.


I dont think we should export the functions to pass to add_stat(), would only export the user facing functions and name them e.g. add_cfr, add_ar and add_crosstab? 
 **-- will remove**
 
 gtsummary_case_fatality_rate() is not happy with a yes/no variable (see code snippet 1 example 1 below) - guessing because is dichotomous but dont know which level to take? 
 **--can set to categorical when creating the gtsummary table, see code below**
 
 
 gtsummary_case_fatality_rate() returns the original died column from the tbl_summary (see code snippet 1 example 2 below) - can we automatically include gt_remove_stat() within the function so that users dont need to add it afterwards? 
**--Yes, will update**

 need to add cfr by strata - currently throws error "cfr by strata is not currently available, ignoring by argument" 
**--Not sure if we should do this - will discuss on a call**


 gtsummary_attack_rate() not cooperating (see code snippet 1 example 3 below) 
**--See code below**

 need to add attack rate by strata - same error as above 
**--Not sure if we should do this - will discuss on a call**

 for merge_gt_univar_counts() not entirely sure its possible to run a proper poisson regression with offset using gtsummary::tbl_uvregression() - cross reference with epiet case studies if unsure how these should be run appropriately
**--Prioritizing mh odds, will table these for now**

```
library("dplyr")
library("gtsummary")

## load data 
linelist <- epidict::gen_data("cholera")

###### EXAMPLE 1 
# epikit::case_fatality_rate_df takes deaths or cases as it's first argument, 
# which must be a logical argument (or an evaluated logical expression)

## make a "died" logical column (1,0 or TRUE, FALSE will both work)
linelist$died <- ifelse(linelist$exit_status %in% c("DH", "DD", "DOA"), 1, 0)

## Run function
linelist %>% select(died) %>% tbl_summary() %>% gtsummary_case_fatality_rate(deaths_var = "died")


####### EXAMPLE 2 
## whereas with a TRUE/FALSE variable it works (TRUE/FALSE version)
linelist$died <- linelist$died == 1
linelist %>% select(died) %>% tbl_summary() %>% gtsummary_case_fatality_rate(deaths_var = died)

######## EXAMPLE 3  

## get count of number dead (different each time generate dataset) 
num_dead <- sum(linelist$died, na.rm = TRUE)

## run epikit attack rate for reference 
epikit::attack_rate(num_dead, 300)

## compare with gtsummary add_ci (not a good option as cant define our own denominator)
linelist %>% 
  select(died) %>% 
  tbl_summary() %>% add_ci()

table(linelist$died)

# when calculating AR by categorical variables, each level of the variable needs 
# a value for population (see examples in the test file for using population
# tables for gender or age)
pop <- c(300, 300)
linelist %>% 
  select(died) %>% 
  # gtsummary handles logicals as dichotomous by default, and only shows on one line, 
  # change to calculate by level by setting the "type" to categorical
  tbl_summary(type = died ~ "categorical") %>% 
  gtsummary_attack_rate(population = pop, 100)
```

**Comment by aspina7** [comment]
just sumarising phonecall: 

1) check if theres an easy workaround to include `by` argument for strata
2) If not then just pull together an example with tbl_merge or left_join with purrr
3) lower priority but move on to the univar_counts stuff (see reprex below as well as [these lines](https://github.com/EPIET/MultivariableAnalysis/blob/master/MVA%202018/5%20poisson%20and%20cox%20reg/Poisson_Cox_tidyverse.Rmd#L375) from EPIET). Nb not clear if possible to pass the appropriate arguments to `method.args` in `tbl_uvregression()`, and on top of that `tbl_regression()` is a different class.... 
4) Move on to mh stuff

> for merge_gt_univar_counts() not entirely sure its possible to run a proper poisson regression with offset using gtsummary::tbl_uvregression() - cross reference with epiet case studies if unsure how these should be run appropriately **--Prioritizing mh odds, will table these for now**

``` r
library(gtsummary) 
library(dplyr)
library(epidict)
    
    linelist <- gen_data("cholera")
    
    linelist$obstime <- as.Date(linelist$date_of_exit) - as.Date(linelist$date_of_consultation_admission) 
    linelist$obstime <- as.numeric(linelist$obstime)
    
    linelist$died <- linelist$exit_status %in% c("DD", "DH", "DOA")
    
    linelist$sex <- linelist$sex == "M"
    
    regression <- glm(died ~ sex, 
                      family = poisson(link = "log"),
                      offset = log(obstime), 
                      data = linelist) 
    
    regression %>% 
      broom::tidy(exponentiate = TRUE, conf.int = TRUE)
    
    
    
    linelist %>% 
      select(sex, died) %>% 
      tbl_uvregression(
        y = died, 
        method = glm, 
        method.args = list(
          family = "poisson",
          # offset = log(obstime)
        )
      )
    
    
    tbl_regression(regression, 
                   exponentiate = TRUE)
```

**Comment by amygimma** [comment]
```{r}
# Alternate to using `by` variable for cfr and attack rate gtsummary wrappers


library("dplyr")
library("gtsummary")

## load data 
linelist <- epidict::gen_data("cholera")

###### EXAMPLE 1 
# epikit::case_fatality_rate_df takes deaths or cases as it's first argument, 
# which must be a logical argument (or an evaluated logical expression)

## make a "died" logical column (1,0 or TRUE, FALSE will both work)
linelist$died <- ifelse(linelist$exit_status %in% c("DH", "DD", "DOA"), 1, 0)


linelist$age_group <- ifelse(linelist$age_years < 18, "Under 18", "Over 18")
linelist$age_group <- ifelse(!is.na(linelist$age_months), "Under 18", linelist$age_group)
table(linelist$sex)
## Run function
gtm <- linelist %>% 
  select(died, sex, age_group) %>% 
  filter(sex == "M") %>%
  tbl_summary(include = age_group, label = age_group ~ "Age group") %>% 
  gtsummary_case_fatality_rate(deaths_var = "died")

gtf <- linelist %>% 
  select(died, sex, age_group) %>% 
  filter(sex == "F") %>%
  tbl_summary(include = age_group, label = age_group ~ "Age group") %>% 
  gtsummary_case_fatality_rate(deaths_var = "died")

gtu <- linelist %>% 
  select(died, sex, age_group) %>% 
  filter(sex == "U") %>%
  tbl_summary(include = age_group, label = age_group ~ "Age group") %>% 
  gtsummary_case_fatality_rate(deaths_var = "died")



gtall <-
  tbl_merge(
    tbls = list(gtm, gtf, gtu),
    tab_spanner = c("Male", "Female", "Unknown")
  )
gtall

```

**Comment by aspina7** [comment]
@alicelouisejane see [this comment](https://github.com/R4EPI/epitabulate/pull/21#issuecomment-1036377060) when get to doing stratified attack rates and cmr in the outbreak templates - as discussed if not a table being shown in the report and just need to produce the numbers can just use the attack_rate function directly or dplyr where necessary. 

@amygimma nice lets just use this for stratified - any idea on when will have time to move on to the univariate / mh stuff? Ah and also importantly: migrating ar and cmr funcitons to here from epikit

**Comment by aspina7** [comment]
@amygimma can you do below and then we can merge please: 

- [ ] rename the exported functions to be just add_* instead of all the gtsummary_ 
- [ ] include mortality_rate() function too 
- [ ] move the functions over from epikit

**Comment by amygimma** [comment]
@alex - this should be good to go as far as the gtsummary add_cfr and add_ar functions go. Can we merge these in and then I'll put up another branch for the remaining stuff?

**Comment by amygimma** [comment]
if this looks ok for cross tabs i can merge this in too, lemme know if you want just numbers and no %:, and can prob also hide the missing values if needed

This is for:
exposure = recent travel
outcome = diarrhoea
variable = gender

Then can add output for mhodds and stuff

Screenshot 2022-03-09 at 03 46 10

**Comment by aspina7** [comment]
> if this looks ok for cross tabs i can merge this in too, lemme know if you want just numbers and no %:, and can prob also hide the missing values if needed
> 
> This is for: exposure = recent travel outcome = diarrhoea variable = gender
> 
> Then can add output for mhodds and stuff

Yes this looks fine for Odds ratios - but remember theres the added complexity of needing different outputs according to the type of estimate (RR and IRR) - as in comment below, and the current epitabulate function for uv regression walks through all that.... 


> 3. lower priority but move on to the univar_counts stuff (see reprex below as well as [these lines](https://github.com/EPIET/MultivariableAnalysis/blob/master/MVA%202018/5%20poisson%20and%20cox%20reg/Poisson_Cox_tidyverse.Rmd#L375) from EPIET). Nb not clear if possible to pass the appropriate arguments to `method.args` in `tbl_uvregression()`, and on top of that `tbl_regression()` is a different class....
> 
> > for merge_gt_univar_counts() not entirely sure its possible to run a proper poisson regression with offset using gtsummary::tbl_uvregression() - cross reference with epiet case studies if unsure how these should be run appropriately **--Prioritizing mh odds, will table these for now**
> 
> ```r
> library(gtsummary) 
> library(dplyr)
> library(epidict)
>     
>     linelist <- gen_data("cholera")
>     
>     linelist$obstime <- as.Date(linelist$date_of_exit) - as.Date(linelist$date_of_consultation_admission) 
>     linelist$obstime <- as.numeric(linelist$obstime)
>     
>     linelist$died <- linelist$exit_status %in% c("DD", "DH", "DOA")
>     
>     linelist$sex <- linelist$sex == "M"
>     
>     regression <- glm(died ~ sex, 
>                       family = poisson(link = "log"),
>                       offset = log(obstime), 
>                       data = linelist) 
>     
>     regression %>% 
>       broom::tidy(exponentiate = TRUE, conf.int = TRUE)
>     
>     
>     
>     linelist %>% 
>       select(sex, died) %>% 
>       tbl_uvregression(
>         y = died, 
>         method = glm, 
>         method.args = list(
>           family = "poisson",
>           # offset = log(obstime)
>         )
>       )
>     
>     
>     tbl_regression(regression, 
>                    exponentiate = TRUE)
> ```

**Review by aspina7 — APPROVED** [review]

**Comment by aspina7** [comment]
Approving this PR and the following are going in new PRs

> * [ ]  rename the exported functions to be just add_* instead of all the gtsummary_
> * [ ]  include mortality_rate() function too
> * [ ]  move the functions over from epikit

**Review by aspina7 — APPROVED** [review]

**Comment by amygimma** [comment]
btw @aspina7 the error here is related to epikit - on next those functions will be moved to this package anyway, so not addressing for now

File: tests/testthat/test-gt_wrappers.R [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+    dplyr::group_by(age_group) %>%
+    dplyr::mutate(total = sum(case_n)) %>%
+    dplyr::filter(case == T)
+
+  # calculate population total from population table
+  expected_ar_lev <- attack_rate(counts$case_n, counts$total, multiplier = 10000) %>%
+    epikit::merge_ci_df(e = 3)
+
+
+  gt_ar_lev <- linelist_cleaned %>%
```
Comment (aspina7) [inline tests/testthat/test-gt_wrappers.R:361]: this errors for me - cant figure out why.... is missing population?? 
```r
Error:
! Dimension of 'age_group' and the added statistic do not match. Expecting statistic/data frame to be length/no. rows 5.
Run `rlang::last_error()` to see where the error occurred.
```

File: tests/testthat/test-gt_wrappers.R [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+    epikit::merge_ci_df(e = 3) %>%
+    mutate(ar = formatC(ar, 2, format = "f"))
+
+  # attack rate for each group
+  expected_ar_lev <- epikit::attack_rate(cases$cases, cases$total, multiplier = 10000) %>%
+    epikit::merge_ci_df(e = 3) %>%
+    dplyr::mutate(cases = as.character(cases)) %>%
+    mutate(ar = formatC(ar, 2, format = "f"))
+
+  gt_ar <- linelist_cleaned %>%
```
Comment (aspina7) [inline tests/testthat/test-gt_wrappers.R:441]: same error here as above for me .... maybe i missed something. tired.

**Comment by aspina7** [comment]
sorry for taking a while to get to this.... getting there - just getting two errors on testing file I cant work out. 

Otherwise: 
- not sure population being used the way intended, need to be able to feed in from a different dataframe (i.e. not in linelist)
- also not sure crosstabs doing what we want to, but maybe im missing something or you are doing it on a separate pr 

can we jump on a call to discuss please? we need to get this wrapped up.

**Comment by aspina7** [comment]
> * also not sure crosstabs doing what we want to, but maybe im missing something or you are doing it on a separate pr

we basically want to be using modify_table_body on (ideally) a gtsummary::tbl_uvregression to add in the correct numbers. 
And that should just be an add_crosstab() wrapper which we can pipe in to from a tbl_uvregression. E.g. what is done [here](https://stackoverflow.com/a/71087530)

**Comment by aspina7** [comment]
@amygimma 
sorry for taking a while to get to this.... getting there - just getting two errors on testing file which I flagged (maybe this is just from what you said because needs to be moved from {epikit}?) 

Otherwise:

- Theres no `population` argument for the `add_ar()` function and there needs to be. I.e. it needs to be possible to add population counts from a separate data.frame (as in code below) - not just summing the linelist rows. 

```r
# define population 
population <- sum(population_data_age$population)
attack_rate(nrow(linelist_cleaned), population, multiplier = 10000)
```
- In `add_mr()` you have allowed for an external population count to feed in but the rate that comes up is wrong (we need to be able to replicate what comes out [here](https://github.com/R4EPI/sitrep/pull/275/files#diff-bcfc7e40f16321bf93c3d603ac54729bf99044ef430b2107d8c375e6893ee9f6R1348))

- for `add_crosstab()`, I was wrong and we can in fact just us inputs from `gtsummary::tbl_uvregression()` (see me being retarded [here](https://github.com/ddsjoberg/gtsummary/issues/1206). So what you have done for cross tabbing is good, but we just want to pull the arguments directly from tbl_uvregression, rather than users building their own table. Remember its not just Odds ratios, we also want for risk ratios and incidence rate ratios - and the counts that we want to display differs for each (replicate what comes out of `epitabulate::tab_univariate()`  (also emailed with some notes for mh). 
    - we basically want to be using modify_table_body on (ideally) a gtsummary::tbl_uvregression to add in the correct numbers. 
And that should just be an add_crosstab() wrapper which we can pipe in to from a tbl_uvregression. E.g. what is done [here](https://stackoverflow.com/a/71087530)

can jump on a call if any of this isn't clear - but clarify if not please because we need to get this wrapped up quick.

**Review by aspina7 — APPROVED** [review]

**Pull request merged**